### DISCUSSION TITLE
Plugging anaconda under LXC zone creation

### DISCUSSION THREAD

**Discussion by kloczek** [rhinstaller/anaconda#4758] [General]
I'm trying to find a way how to use kickkstart to install LXC zone.
To be able use kickstart file to install LC zone it is necessary to answer on below questions:
- what needs to be executed to init kikstart process with exact .ks file?
- how to disable in .ks file boot loader setup?
- how to disable storage initialisation for root fs?

I would be appreciated if someone will be able to find answers o above questions and/or provide any suggestions how to approach to use anaconda on implementing LXC zone install.

**Comment by VladimirSlavik** [comment]
I am not sure if I get the question completely, but that might be because I don't know too much about LXC. I'll make the reply longer to reasonably cover what I think is relevant - sorry if you have to pick the right bits from it.

---

First, just straight answers to the questions...

> what needs to be executed to init kikstart process with exact .ks file?

Anaconda must start, have a complete kickstart, and a viable installation target.

> how to disable in .ks file boot loader setup?

I think `bootloader --disabled`.

> how to disable storage initialisation for root fs?

You need to run the installation in dirinstall mode. This is a bit tricky, as there are few references to it, because there is no documentation; check here for the tiny bits of info: https://github.com/rhinstaller/anaconda/blob/master/data/anaconda_options.txt Essentially you start anaconda with the `--dirinstall` option, it reads the `ANACONDA_ROOT_PATH` variable, and if that is missing, defaults to `/mnt/sysimage`. Dirinstall disables all storage work, as well as the bootloader. I don't know further details myself - searching the code for `conf.target.is_directory` should provide more clues as to what is different.

---

Second, the more open ended part.

I am not clear on how you plan to to install "into" a container. Can you somehow import an archive, image, or something to create the LXC container? That's how Fedora does it. They use anaconda with kickstart to create the base image. There was an excellent article on the Fedora blog some time ago that describes some of that: https://fedoramagazine.org/how-to-build-fedora-container-images/ The article mostly leaves out the crucial part "where does the archive come from": There's only a vague description and a dead link to Koji, but you could probably swap in a newer compose date into the url and it should work. Anyway, connecting with the [Fedora infra?] people doing that left-out part might help you far more than I can.

There's also https://fedoramagazine.org/build-smaller-containers/ where the "Building containers from scratch" part might be relevant.

Finally, just to make you aware, beside dirinstall, there is also image installation. I never had to care about either, so I can't tell you which is the correct one you will need for your project.

**Reply by kloczek** [reply]
> Second, the more open ended part.
> 
> I am not clear on how you plan to to install "into" a container. Can you somehow import an archive, image, or something to create the LXC container?

Process is very simillar to install zone on Solaris or *BSD in jails:
- zone creaion process prepares storage and empty direcory wrehere system image will be assembled
- using list of packages it executes package management software to install list of packages in that already prapred root directory
- after installtion are taken additional steps like setup and/or update zone boot management data (on Solatis it is possible to have on zone boot grub like prompt to choose one of the instances of cloned root file system)

In case of Linux it is possible to use dnf with `--installroot` howver I want to have zone installation process as kicks start process to be able to reuse exacly the same file on bare metal installation or reuse bar metakl install ks to install system in zone (that is the goal).

> There's also https://fedoramagazine.org/build-smaller-containers/ where the "Building containers from scratch" part might be relevant.

No I want to be able to use kickstart to deploy zone. This page has noting to do with kickstart and want to reuse it on zone deployment.

Other thing is that if at some point it will be working I would like to extend kickstart to be able specyfy separated .ks files to deploy additional zones  on bare metal installation like it is on Solarizs where additional AI (Automated Installer) manifests (xml description) can be specifid to automatic zones creation and setup on bare matal installation.

You can peak on how it is done on Solaris https://docs.oracle.com/cd/E23824_01/html/E21798/aizoneinst.html

**Reply by kloczek** [reply]
> You need to run the installation in dirinstall mode. This is a bit tricky, as there are few references to it, because there is no documentation; check here for the tiny bits of info: https://github.com/rhinstaller/anaconda/blob/master/data/anaconda_options.txt Essentially you start anaconda with the `--dirinstall` option, it reads the `ANACONDA_ROOT_PATH` variable, and if that is missing, defaults to `/mnt/sysimage`. Dirinstall disables all storage work, as well as the bootloader. I don't know further details myself - searching the code for `conf.target.is_directory` should provide more clues as to what is different.

So what exactly needs to be executed on the system with installed anaconda to install everything what is described if I have in lets say /etc/zonec/zone1.ks file and empty zone root fs in /var/lib/lxc/zone1/rootfs? 🤔

**Reply by VladimirSlavik** [reply]
I'd guess something like
```
ANACONDA_ROOT_PATH=/var/lib/lxc/zone1/rootfs anaconda --dirinstall --kickstart=/etc/zonec/zone1.ks
```
- You might need to link the kickstart as file:// or serve from http - not sure.
- You don't want to play with this on your machine, it's a fast way to discover anaconda can indeed format disks.
- Also, note that dirinstall isn't really supported as such.

At any rate, you're hitting the limits of my knowledge here, I don't think I can help more.

You might also ask Image builder, they are either gearing towards something similar, or doing it already.

**Comment by Chewbakka-Wakka** [comment]
I too share the same requirements as previously stated and I've been both searching around while also doing a touch of reverse engineering to see how the current images are built and uploaded. The current process uses higher level tools therefore a greater set of dependencies and thus complexity to implement, using anaconda at a lower level offers a few advantages.

* Reuse of exactly the same file(s) on bare metal to deploy one or more lxc/zones
* Fewer dependencies than other methods
* Extending Kickstart to be able specify separated .ks files to deploy additional zones on bare metal installation 

My thoughts on these points during the interim as a POC would effectively be nesting another call to anaconda within the %post section calling those very commands and using an include file containing the lxc based ks profile. Image installation can follow once proven that this works.

Also a new package group could be used/created such as "@lxc" much like "@minimal" granted lxc/zones require less package as on the host OS to function.

On a test box in the meantime...

Syntax is correct, slight change:
`[root@te ~]# lxc-create -t none -n fedora-test -B btrfs -- -d fedora -r 38 -a amd64
[root@te ~]# mkdir /var/lib/lxc/fedora-test/rootfs
[root@te ~]# ANACONDA_ROOT_PATH=/var/lib/lxc/fedora-test/rootfs anaconda --text --debug --dirinstall --kickstart=/root/ks/lxc/lxc.ks 
Starting installer, one moment...
** (process:678): WARNING **: 10:29:26.418: failed to load module lvm: libbd_lvm.so.2: cannot open shared object file: No such file or directory

** (process:678): WARNING **: 10:29:26.418: failed to load module btrfs: libbd_btrfs.so.2: cannot open shared object file: No such file or directory

** (process:678): WARNING **: 10:29:26.418: failed to load module swap: libbd_swap.so.2: cannot open shared object file: No such file or directory

** (process:678): WARNING **: 10:29:26.419: failed to load module loop: libbd_loop.so.2: cannot open shared object file: No such file or directory

** (process:678): WARNING **: 10:29:26.419: failed to load module crypto: libbd_crypto.so.2: cannot open shared object file: No such file or directory

** (process:678): WARNING **: 10:29:26.419: failed to load module mpath: libbd_mpath.so.2: cannot open shared object file: No such file or directory

** (process:678): WARNING **: 10:29:26.419: failed to load module dm: libbd_dm.so.2: cannot open shared object file: No such file or directory

** (process:678): WARNING **: 10:29:26.419: failed to load module mdraid: libbd_mdraid.so.2: cannot open shared object file: No such file or directory

** (process:678): WARNING **: 10:29:26.419: failed to load module nvdimm: libbd_nvdimm.so.2: cannot open shared object file: No such file or directory
10:29:27 Kernel boot argument 'selinux' detected. Did you want to use 'inst.selinux' for the installer instead?
10:29:27 All Anaconda kernel boot arguments are now required to use 'inst.' prefix!
anaconda 38.23.4-2.fc38 for anaconda bluesky (pre-release) started`

This does work however upon starting the lxc/zone once install finishes/exits I see this:
`[root@te lxc]# lxc-start -n fedora-test -F --logfile /tmp/lxc.log --logpriority TRACE`

`[root@te lxc]# grep -i err /tmp/lxc.log
lxc-start fedora-test 20230802223236.549 ERROR    conf - conf.c:lxc_transient_proc:3809 - No such file or directory - Failed to mount temporary procfs
lxc-start fedora-test 20230802223236.549 ERROR    conf - conf.c:bind_mount_console:1942 - No such file or directory - Failed to open "/dev/pts/1"`

Looks like 3 interacting parts between systemd, lxc and anaconda.

` [root@te lxc]# rpm -qa systemd\*
systemd-libs-253.7-1.fc38.x86_64
systemd-pam-253.7-1.fc38.x86_64
systemd-resolved-253.7-1.fc38.x86_64
systemd-253.7-1.fc38.x86_64
systemd-udev-253.7-1.fc38.x86_64
systemd-boot-unsigned-253.7-1.fc38.x86_64

[root@te lxc]# rpm -qa anaconda\*
anaconda-tui-38.23.4-2.fc38.x86_64
anaconda-core-38.23.4-2.fc38.x86_64

[root@te lxc]# rpm -qa lxc\*
lxc-libs-4.0.12-2.fc37.x86_64
lxc-4.0.12-2.fc37.x86_64
lxc-templates-4.0.12-2.fc37.x86_64`
